core/tee/tee_time_generic.c

```diff
@@ -17,50 +21,179 @@ struct tee_ta_time_offs {
+		tee_svc_storage_file_ops(TEE_STORAGE_PRIVATE);
+	struct tee_file_handle *fh = NULL;
+	struct tee_pobj *po = NULL;
+	struct tee_ta_time_offs o = { };
+	TEE_Result res = TEE_SUCCESS;
+
+	if (!fops)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	res = tee_pobj_get((TEE_UUID *)uuid, (void *)TA_TIME_OFFS_ID,
+			   sizeof(TA_TIME_OFFS_ID) - 1,
+			   TEE_DATA_FLAG_ACCESS_WRITE,
+			   TEE_POBJ_USAGE_CREATE, fops, &po);
+
+	if (res)
+		return res;
+
+	o.uuid = *uuid;
+	o.offs = *offs;
+	o.positive = positive;
+
+	res = fops->create(po, true, NULL, 0, NULL, 0, &o, NULL, sizeof(o),
+		&fh);
+
+	if (!res) {
+		fops->close(&fh);
+		tee_pobj_create_final(po);
+	}
+	tee_pobj_release(po);
+	return res;
+}
+
+static TEE_Result tee_time_ta_storage_read(const TEE_UUID *uuid,
+					   TEE_Time *offs, bool *positive)
+{
+	const struct tee_file_operations *fops =
+		tee_svc_storage_file_ops(TEE_STORAGE_PRIVATE);
+	struct tee_file_handle *fh = NULL;
+	struct tee_pobj *po = NULL;
+	TEE_Result res = TEE_SUCCESS;
+	size_t sz = 0;
+	struct tee_ta_time_offs o = { };
+
+	if (!fops)
+		return TEE_ERROR_TIME_NOT_SET;
+
+	res = tee_pobj_get((TEE_UUID *)uuid, (void *)TA_TIME_OFFS_ID,
+			   sizeof(TA_TIME_OFFS_ID) - 1,
+			   TEE_DATA_FLAG_ACCESS_READ, TEE_POBJ_USAGE_OPEN,
+			   fops, &po);
+
+	if (res) {
+		if (res == TEE_ERROR_ITEM_NOT_FOUND)
+			return TEE_ERROR_TIME_NOT_SET;
+		return res;
+	}
+
+	res = fops->open(po, &sz, &fh);
+	if (!res) {
+		sz = sizeof(o);
+		res = fops->read(fh, 0, &o, NULL, &sz);
+		fops->close(&fh);
+	}
+
+	tee_pobj_release(po);
+
+	if (res) {
+		if (res == TEE_ERROR_ITEM_NOT_FOUND)
+			return TEE_ERROR_TIME_NOT_SET;
+		return res;
+	}
+
+	if (sz != sizeof(o))
+		return TEE_ERROR_TIME_NOT_SET;
+
+	*offs = o.offs;
+	*positive = o.positive;
+	return TEE_SUCCESS;
+}
+
+static TEE_Result tee_time_ta_get_offs(const TEE_UUID *uuid,
+				       const TEE_Time **offs, bool *positive)
+{
+	size_t n = 0;
+	TEE_Time o;
+	bool pos = false;
+	size_t idx = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	mutex_lock(&tee_time_mtx);
+	for (n = 0; n < tee_time_num_offs; n++) {
+		if (memcmp(uuid, &tee_time_offs[n].uuid, sizeof(TEE_UUID))
+				== 0) {
+			*offs = &tee_time_offs[n].offs;
+			*positive = tee_time_offs[n].positive;
+			mutex_unlock(&tee_time_mtx);
+			return TEE_SUCCESS;
+		}
+	}
+
+	mutex_unlock(&tee_time_mtx);
+	res = tee_time_ta_storage_read(uuid, &o, &pos);
+
+	if (res)
+		return res;
+
+	res = tee_time_ta_set_offs_mem(uuid, &o, pos, &idx);
+	if (res)
+		return res;
+
+	mutex_lock(&tee_time_mtx);
+	*offs = &tee_time_offs[idx].offs;
+	*positive = tee_time_offs[idx].positive;
+	mutex_unlock(&tee_time_mtx);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result tee_time_ta_set_offs(const TEE_UUID *uuid,
+				       const TEE_Time *offs, bool positive)
+{
+	size_t idx = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	res = tee_time_ta_set_offs_mem(uuid, offs, positive, &idx);
+	if (res)
+		return res;
+	return tee_time_ta_storage_write(uuid, offs, positive);
```

Comment:
What if `tee_time_ta_storage_write()` fails? Then we have an inconsistent state.